beard/similarity/fuzzy.py

```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Beard.
+# Copyright (C) 2015 CERN.
+#
+# Beard is a free software; you can redistribute it and/or modify it
+# under the terms of the Revised BSD License; see LICENSE file for
+# more details.
+
+"""Fuzzy Similarity for paired data.
+
+.. codeauthor:: Evangelos Tzemis <evangelos.tzemis@cern.ch>
+
+"""
+
+import numpy as np
+import scipy.sparse as sp
+import operator
+
+from sklearn.base import TransformerMixin
+
+
+class FuzzySimilarity(TransformerMixin):
+
+    """Implements Fuzzy Similarity (Soft TF-IDF) metric."""
+
+    def __init__(self, vocabulary, distance, threshold):
+        """Initialize.
+
+        Parameters
+        ----------
+        :param vocabulary:
+            From word to feature_id or opposite.
+
+        :param distance:
+            Distance function which returns anumper in R.
+
+        :param threshold:
+            Threshold parameter that belongs to R.
+        """
+        self.vocabulary = vocabulary
+        self.distance = distance
+        self.threshold = threshold
+
+    def fit(self, X, y=None):
+        """Return self."""
+        return self
+
+    def transform(self, X):
+        """Compute the fuzzy similarity for all pairs of elements in `X`.
+
+        Rows i in `X` are assumed to represent pairs, where
+        `X[i, :n_features]` and `X[i, n_features:]` correspond to their two
+        individual elements. Calling `transform` computes the fuzzy similarity
+        between these elements.
+
+        Parameters
+        ----------
+        :param X: array-like, shape (n_samples, 2 * n_features)
+            Input paired data.
+
+        Returns
+        -------
+        :returns X_: array-like, shape (n_samples, 2 * n_features_prime)
+            The transformed data.
+        """
+        n_samples, n_features_all = X.shape
+        n_features = n_features_all / 2
+
+        if sp.issparse(X):
+            X = X.todense()
+
+        X1 = X[:, :n_features]
+        X2 = X[:, n_features:]
+
+        summ = 0
+        # for each document in X1
+        for row, u in enumerate(X1):
+            # for each word in document u
+            for pos, w in enumerate(u):
+                if w > 0.0:
+                    w_word = self.vocabulary[pos]
+                    min_d, n_index = self._find_distance(w_word, X2[row, :])
+                    summ += w * X2[row, n_index] * min_d
```

Comment:
I am not sure to understand. So you are basically computing a weighted dot product. I think it would be better it this was normalized, to have something close to `CosineSimilarity`.